### PR TITLE
[Wasi] switch to preview 2

### PR DESCRIPTION
**Pull request by pavelsavara**
After this PR dotnet will **produce WASM component (preview 2)** instead of WASM module
- it has different binary header
- component consumes WASI imports
- it's not compatible with wasm-opt and old versions of wasm engines

### Details
- switch to WASIp2 sysroot
- temporary `wasi-sdk-p2.cmake` file 
    - to be removed when WASI SDK 23 lands
    - see https://github.com/WebAssembly/wasi-sdk/pull/430
- new environment variable `DOTNET_WASI_PRINT_EXIT_CODE`
    - will print `WASM EXIT 42` into stderr when it's not zero
    - because WASI components don't return [numeric exit code](https://github.com/WebAssembly/wasi-cli/issues/11)
    - and because our unit tests usually need to know more than 0/1
- new `dotnet run` option `--forward-exit-code` which sets it
   - the dotnet host will consume the line from stderr and return exit code to caller shell
- minor wasi cleanup
- fail fast if there is `wasm-opt` on PATH

### WASI SDK 22 + patches
- force `WASI_SDK_PATH` to `src\mono\wasi\wasi-sdk` if the env `WASI_SDK_PATH` points to unpatched location
- new `VERSION22PATCHED` file to recognize patched installation
- new `_BuildNativeEnvironmentVariables` for building native libs to propagate correct `WASI_SDK_PATH` to shell
Most of it could go away after update to WASI SDK 23. see https://github.com/dotnet/runtime/issues/104773

Contributes to https://github.com/dotnet/runtime/issues/96419
Split from https://github.com/dotnet/runtime/pull/103752

### CODE REVIEW

File: src/mono/wasi/runtime/main.c [hunk 1]
```diff
@@ -110,6 +111,14 @@ int main(int argc, char * argv[]) {
 		mono_print_unhandled_exception(out_exc);
 		exit(1);
 	}
-	return ret < 0 ? -ret : ret;
+	ret = ret < 0 ? -ret : ret;
+
+	char* dotnet_wasi_print_exit_code = monoeg_g_getenv ("DOTNET_WASI_PRINT_EXIT_CODE");
```
Comment (pavelsavara) [inline src/mono/wasi/runtime/main.c:116]: @dicej @SingleAccretion 

I'm introducing this trick to be able to work with proper exit code in various tests we have.
maybe NativeAOT could do the same.
Feedback welcome.

Comment (pavelsavara) [reply src/mono/wasi/runtime/main.c:116]: We have similar logic in Browser

Comment (SingleAccretion) [reply src/mono/wasi/runtime/main.c:116]: > new dotnet run option --forward-exit-code which sets it

This is _not_ the default for `run`, correct?

To be honest, it is sad to see this workaround being needed, but it is what it is.

Comment (pavelsavara) [reply src/mono/wasi/runtime/main.c:116]: no, it's not a default

**Comment by pavelsavara** [comment]
cc @dicej, feedback welcome

File: eng/native/gen-buildsys.cmd [hunk 2]
```diff
@@ -73,7 +74,7 @@ if /i "%__Arch%" == "wasm" (
         set "WASI_SDK_PATH=!WASI_SDK_PATH:\=/!"
         if not "!WASI_SDK_PATH:~-1!" == "/" set "WASI_SDK_PATH=!WASI_SDK_PATH!/"
         set __CmakeGenerator=Ninja
-        set __ExtraCmakeParams=%__ExtraCmakeParams% -DCLR_CMAKE_TARGET_OS=wasi -DCLR_CMAKE_TARGET_ARCH=wasm "-DWASI_SDK_PREFIX=!WASI_SDK_PATH!" "-DCMAKE_TOOLCHAIN_FILE=!WASI_SDK_PATH!share/cmake/wasi-sdk.cmake" "-DCMAKE_SYSROOT=!WASI_SDK_PATH!share/wasi-sysroot"
+        set __ExtraCmakeParams=%__ExtraCmakeParams% -DCLR_CMAKE_TARGET_OS=wasi -DCLR_CMAKE_TARGET_ARCH=wasm "-DWASI_SDK_PREFIX=!WASI_SDK_PATH!" "-DCMAKE_TOOLCHAIN_FILE=!__repoRoot!/src/native/external/wasi-sdk-p2.cmake" "-DCMAKE_SYSROOT=!WASI_SDK_PATH!share/wasi-sysroot" "-DCMAKE_CROSSCOMPILING_EMULATOR=!EMSDK_NODE! --experimental-wasm-bigint --experimental-wasi-unstable-preview1"
```
Comment (dicej) [inline eng/native/gen-buildsys.cmd:77]: I suspect this `EMSDK_NODE` reference is the last remaining dependency on Emscripten when building for WASI.  It seems fine for CI, where we install the Emscripten SDK anyway, but it would be nice to be able to build locally without installing it.

I guess folks could just set `EMSDK_NODE=node` in their environment, and we could document that as a legitimate option.  Perhaps we could also rename it to e.g. `NODE_EXECUTABLE` to make it clear that there's nothing Emscripten-specific about the version of Node required?

File: eng/native/gen-buildsys.sh [hunk 3]
```diff
@@ -98,7 +98,7 @@ if [[ "$host_arch" == "wasm" ]]; then
     if [[ "$target_os" == "browser" ]]; then
         cmake_command="emcmake $cmake_command"
     elif [[ "$target_os" == "wasi" ]]; then
-        cmake_extra_defines="$cmake_extra_defines -DCLR_CMAKE_TARGET_OS=wasi -DCLR_CMAKE_TARGET_ARCH=wasm -DWASI_SDK_PREFIX=$WASI_SDK_PATH -DCMAKE_TOOLCHAIN_FILE=${WASI_SDK_PATH}share/cmake/wasi-sdk.cmake" "-DCMAKE_SYSROOT=${WASI_SDK_PATH}share/wasi-sysroot"
+        cmake_extra_defines=("$cmake_extra_defines" "-DCLR_CMAKE_TARGET_OS=wasi" "-DCLR_CMAKE_TARGET_ARCH=wasm" "-DWASI_SDK_PREFIX=$WASI_SDK_PATH" "-DCMAKE_TOOLCHAIN_FILE=$WASI_SDK_PATH/share/cmake/wasi-sdk.cmake" "-DCMAKE_SYSROOT=${WASI_SDK_PATH}share/wasi-sysroot" "-DCMAKE_CROSSCOMPILING_EMULATOR=node --experimental-wasm-bigint --experimental-wasi-unstable-preview1")
```
Comment (dicej) [inline eng/native/gen-buildsys.sh:101]: Would you mind cherry-picking https://github.com/dotnet/runtimelab/pull/2614/commits/b2e8ff4819c4b2a5564b34409704b290610eab7f, since this won't work without that patch?

Also, per my comment on `gen-buildsys.cmd`, we should probably be consistent about how we select which `node` to use (i.e. decide whether to just always use whatever's in PATH or make it an environment variable).

File: src/mono/mono.proj [hunk 4]
```diff
@@ -358,6 +358,16 @@ JS_ENGINES = [NODE_JS]
     </PropertyGroup>
     <Error Text="Expected version: %(_ExpectedVersionLines.Identity) and actual version: %(_ActualVersionLines.Identity) of WASI SDK does not match. Please delete $(WASI_SDK_PATH) folder to provision a new version."
       Condition="'$(ActualWasiSdkVersion)' != '$(ExpectedWasiSdkVersion)'" />
+
+    <!-- LLVM in WASI SDK 22 will call wasm-opt when found on the PATH. But it will fail because wasm-opt can't read the WASM components.
+    After we upgrade to WASI SDK 23, we could use no-wasm-opt LLVM option to avoid this issue.
```
Comment (SingleAccretion) [inline src/mono/mono.proj:363]: Open an issue about this? Possibly a generic "tasks to do when upgrading to WASI SDK 23". It would include removing `wasi-sdk-p2.cmake`.

File: src/mono/wasi/build/WasiApp.InTree.targets [hunk 5]
```diff
@@ -10,13 +10,7 @@
...
   <!-- FIXME: use proper dependency -->
   <Target Name="SetupAppHostConfig" BeforeTargets="_WasmGenerateAppBundle">
-    <PropertyGroup>
-      <_WasmMainJSFileName>$([System.IO.Path]::GetFileName('$(WasmMainJSPath)'))</_WasmMainJSFileName>
-    </PropertyGroup>
     <ItemGroup>
-        <HostConfig Include="v8" Host="v8" js-path="$(_WasmMainJSFileName)" />
-        <HostConfig Include="node" Host="nodejs" js-path="$(_WasmMainJSFileName)" />
-        <HostConfig Include="browser" Host="browser" html-path="index.html" />
     </ItemGroup>
```
Comment (SingleAccretion) [inline src/mono/wasi/build/WasiApp.InTree.targets:14]: Nit: empty `ItemGroup`.

File: src/mono/wasi/build/WasiApp.targets [hunk 6]
```diff
@@ -185,6 +185,7 @@
       <_WasiClangCommonFlags Include="$(_DefaultWasiClangFlags)" />
       <_WasiClangCommonFlags Include="$(WasiClangFlags)" />
       <_WasiClangCommonFlags Include="--sysroot=&quot;$(WASI_SDK_PATH.Replace('\', '/'))share/wasi-sysroot&quot;" />
+      <_WasiClangCommonFlags Include="--target=wasm32-unknown-wasip2" />
```
Comment (SingleAccretion) [inline src/mono/wasi/build/WasiApp.targets:188]: Meta-note: there are a lot of places where the target triple is duplicated.

**Review by dicej — COMMENTED** [review]
LGTM overall.  Given that some of the Bash files have changed, is one of the goals of this PR to support Linux and/or MacOS?  If so, I can try building and running the smoke tests locally.  Otherwise, I can fix any problems with the Linux build that pop up downstream when rebasing and testing https://github.com/dotnet/runtimelab/pull/2614.

File: eng/native/gen-buildsys.cmd (hunk 2 above)
Comment (SingleAccretion) [reply eng/native/gen-buildsys.cmd:77]: `CMAKE_CROSSCOMPILING_EMULATOR` is actually a downstream-only need. We can:

1) Delete it here, and add downstream.
2) Delete it here, and hardcode the relevant cmake test results in `tryrun.cmake`.
2) Leave it here as just `node` (from `PATH`).

My slight preference is option `3`.

Comment (pavelsavara) [reply eng/native/gen-buildsys.cmd:77]: the unix version also had just 3), updated

File: eng/native/gen-buildsys.sh (hunk 3 above)
Comment (pavelsavara) [reply eng/native/gen-buildsys.sh:101]: this is bad merge. I wanted p2 tool file, but not the extra `cmake_extra_defines_wasm`, let me try again.

File: src/mono/mono.proj (hunk 4 above)
Comment (pavelsavara) [reply src/mono/mono.proj:363]: https://github.com/dotnet/runtime/issues/104773

File: src/mono/wasi/build/WasiApp.targets (hunk 6 above)
Comment (pavelsavara) [reply src/mono/wasi/build/WasiApp.targets:188]: They don't really share nice place where we could have a constant.

Comment (pavelsavara) [reply src/mono/wasi/build/WasiApp.targets:188]: and maybe effectively, this is a constant :)

File: eng/native/gen-buildsys.sh (hunk 3 above)
Comment (pavelsavara) [reply eng/native/gen-buildsys.sh]: ok, I don't know how to do it without `cmake_extra_defines_wasm`, I take your version

File: src/native/libs/build-native.proj [hunk 7]
```diff
@@ -26,6 +26,11 @@
     <ProjectReference Include="$(LibrariesProjectRoot)native-binplace.proj" />
   </ItemGroup>
 
+  <PropertyGroup Condition="'$(TargetOS)' == 'wasi'">
+    <WASI_SDK_PATH Condition="'$(WASI_SDK_PATH)' == '' or !Exists('$(WASI_SDK_PATH)/VERSION')">$([MSBuild]::NormalizeDirectory($(RepoRoot), 'src', 'mono', 'wasi', 'wasi-sdk'))</WASI_SDK_PATH>
```
Comment (pavelsavara) [inline src/native/libs/build-native.proj:30]: This does not propagate to the linux shell for me and uses original `WASI_SDK_PATH` env variable set to too old SDK. I need to improve this.

File: src/mono/mono.proj [hunk 8]
```diff
@@ -339,8 +339,14 @@ JS_ENGINES = [NODE_JS]
           WorkingDirectory="$(ArtifactsObjDir)"
           IgnoreStandardErrorWarningFormat="true" />
 
-    <Copy SourceFiles="$(MSBuildThisFileDirectory)/wasi/wasi-sdk/share/wasi-sysroot/include/wasm32-wasi-threads/pthread.h"
-          DestinationFiles="$(MSBuildThisFileDirectory)/wasi/wasi-sdk/share/wasi-sysroot/include/wasm32-wasip2/pthread.h"
+    <!-- Temporary in WASI-SDK 22
+    TODO https://github.com/dotnet/runtime/issues/104773
+    -->
+    <Copy SourceFiles="$(WASI_SDK_PATH)/share/wasi-sysroot/include/wasm32-wasi-threads/pthread.h"
```
Comment (pavelsavara) [inline src/mono/mono.proj:345]: If `WASI_SDK_PATH` is set by the CI agent to some read-only location, this will fail together with other patches we do here. I need to re-consider.

Comment (pavelsavara) [reply src/mono/mono.proj:345]: Actually the provisioning and patching will be skipped.

Comment (pavelsavara) [reply src/mono/mono.proj:345]: solved by adding `$(WASI_SDK_PATH)/VERSION22PATCHED` stamp file and checking it everywhere. Also to decide if we need to do provisioning

File: src/native/libs/build-native.proj (hunk 7 above)
Comment (pavelsavara) [reply src/native/libs/build-native.proj]: solved by adding `_BuildNativeEnvironmentVariables`

**Comment by pavelsavara** [comment]
There is https://github.com/dotnet/runtime/issues/104786

**Review by maraf — APPROVED** [review]
Changes looks good to me 👍

**Comment by pavelsavara** [comment]
/ba-g last still running test is not relevant for this change, other CI issues are known and matched

**Comment by dicej** [comment]
> @dicej since I'm out next week, I will try to merge as much as possible today. I will be happy to help to upstream any of your fixes later.

Sounds good.  I'll be out the following week, but we can sync up after that.  Enjoy your time off!

**Comment by SingleAccretion** [comment]
> Overall I think that we should consider broader set of changes from the fork. In order to make future code flow downstream easier. @SingleAccretion are there such candidates ?

I don't think there is a lot that can be upstream (and we try to keep it that way). As you've seen in the merge, the vast majority of the actual NAOT-LLVM code is concentrated in the Jit/ILC/NAOT runtime.

One thing that I would love to unify is the EMSDK layout between dotnet/emsdk and upstream EMSDK. But that's orthogonal to this WASI effort (we are, fortunately, aligned on the build system with WASI).
